packages/fxa-content-server/app/scripts/lib/auth/crypto.ts

```diff
@@ -0,0 +1,179 @@
+      info: encoder().encode(`${NAMESPACE}authPW`),
+      hash: 'SHA-256',
+    },
+    quickStretchedKey,
+    256
+  );
+  const unwrapBKey = await crypto.subtle.deriveBits(
+    {
+      name: 'HKDF',
+      salt: new Uint8Array(0),
+      // @ts-ignore
+      info: encoder().encode(`${NAMESPACE}unwrapBkey`),
+      hash: 'SHA-256',
+    },
+    quickStretchedKey,
+    256
+  );
+  return {
+    authPW: uint8ToHex(new Uint8Array(authPW)),
+    unwrapBKey: uint8ToHex(new Uint8Array(unwrapBKey)),
+  };
+}
+
+export async function deriveBundleKeys(
+  key: hexstring,
+  keyInfo: string,
+  payloadBytes: number = 64
+) {
+  const baseKey = await crypto.subtle.importKey(
+    'raw',
+    hexToUint8(key),
+    'HKDF',
+    false,
+    ['deriveBits']
+  );
+  const keyMaterial = await crypto.subtle.deriveBits(
+    {
+      name: 'HKDF',
+      salt: new Uint8Array(0),
+      // @ts-ignore
+      info: encoder().encode(`${NAMESPACE}${keyInfo}`),
+      hash: 'SHA-256',
+    },
+    baseKey,
+    (32 + payloadBytes) * 8
+  );
+  const hmacKey = await crypto.subtle.importKey(
+    'raw',
+    new Uint8Array(keyMaterial, 0, 32),
+    {
+      name: 'HMAC',
+      hash: 'SHA-256',
+      length: 256,
+    },
+    true,
+    ['verify']
+  );
+  const xorKey = new Uint8Array(keyMaterial, 32);
+  return {
+    hmacKey,
+    xorKey,
+  };
+}
+
+export async function unbundleKeyFetchResponse(
+  key: hexstring,
+  bundle: hexstring
+) {
+  const b = hexToUint8(bundle);
+  const keys = await deriveBundleKeys(key, 'account/keys');
+  const ciphertext = b.subarray(0, 64);
+  const expectedHmac = b.subarray(b.byteLength - 32);
+  const valid = await crypto.subtle.verify(
+    'HMAC',
+    keys.hmacKey,
+    expectedHmac,
+    ciphertext
+  );
+  if (!valid) {
+    throw new Error('Bad HMac');
+  }
+  const keyAWrapB = xor(ciphertext, keys.xorKey);
+  return {
+    kA: uint8ToHex(keyAWrapB.subarray(0, 32)),
+    wrapKB: uint8ToHex(keyAWrapB.subarray(32)),
+  };
+}
+
+export function unwrapKB(wrapKB: hexstring, unwrapBKey: hexstring) {
+  return uint8ToHex(xor(hexToUint8(wrapKB), hexToUint8(unwrapBKey)));
+}
+
+export async function checkWebCrypto() {
+  try {
+    await crypto.subtle.importKey(
+      'raw',
+      crypto.getRandomValues(new Uint8Array(16)),
+      'PBKDF2',
+      false,
+      ['deriveKey']
+    );
+    await crypto.subtle.importKey(
+      'raw',
+      crypto.getRandomValues(new Uint8Array(16)),
+      'HKDF',
+      false,
+      ['deriveKey']
+    );
+    await crypto.subtle.importKey(
+      'raw',
+      crypto.getRandomValues(new Uint8Array(16)),
+      {
+        name: 'HMAC',
+        hash: 'SHA-256',
+        length: 256,
+      },
+      false,
+      ['sign']
+    );
+    await crypto.subtle.digest(
+      'SHA-256',
+      crypto.getRandomValues(new Uint8Array(16))
+    );
+    return true;
+  } catch (err) {
+    try {
+      console.warn('loading webcrypto shim');
+      // prettier-ignore
+      // @ts-ignore
```

Comment:
ignore the fact that `window` doesn't already have an `asmCrypto` property

Reply:
The other way to do this is to cast to any:
```
(window as any).asmCrypt =....
```